### ISSUE TITLE
azurerm_kubernetes_cluster with user assigned managed identity does not assign correctly

### ISSUE THREAD

**Issue by tehho** [terraform-providers/terraform-provider-azurerm#10938]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform: 0.14.7
AzureRM: 2.50.0

### Affected Resource(s)

* `azurerm_kubernetes_cluster`

### Terraform Configuration Files

```hcl
data "azurerm_user_assigned_identity" "identity" {
  name                = local.msi_name
  resource_group_name = data.azurerm_resource_group.group.name
}

data "azurerm_resource_group" "group" {
  name     = local.resource_group_name
}

resource "azurerm_kubernetes_cluster" "aks" {
  name                = local.cluster_name
  location            = local.location
  resource_group_name = data.azurerm_resource_group.group.name
  dns_prefix          = local.cluster_name

  kubernetes_version  = var.cluster_version
  node_resource_group = local.resource_group_name_resources

  default_node_pool {
    name                = local.system_pool.name
    vm_size             = local.system_pool.vm_size
    enable_auto_scaling = true
    max_count           = local.system_pool.max_count
    min_count           = local.system_pool.min_count
    max_pods            = local.system_pool.max_pods
    os_disk_size_gb     = local.system_pool.os_disk_size_gb
    type                = "VirtualMachineScaleSets"
    vnet_subnet_id      = var.subnet.id
    availability_zones  = local.system_pool.availability_zones
  }

  identity {
      type = "UserAssigned"
      user_assigned_identity_id = data.azurerm_user_assigned_identity.identity.user_assigned_identity_id
  }

  role_based_access_control {
    enabled = true

    azure_active_directory {
      managed                = true
      admin_group_object_ids = var.admin_group_object_ids
      tenant_id              = var.tenant_id
    }
  }

  api_server_authorized_ip_ranges = local.api_server_authorized_ip_ranges
  
  addon_profile {
    oms_agent {
      enabled                    = var.log_analytics_workspace.id != "" && var.log_analytics_workspace.id != null
      log_analytics_workspace_id = var.log_analytics_workspace.id == "" ? null : var.log_analytics_workspace.id
    }
  }

  network_profile {
    network_plugin = "azure"
    network_policy = "calico"

    load_balancer_sku = "Standard"

    load_balancer_profile {
      outbound_ip_address_ids = [var.pip_outbound.id]
    }
  }

  tags = var.tags
}

```

### Expected Behaviour

The msi should be associated on the vmss

### Actual Behaviour

Creates a new msi under the aks folder and associates that instead.

### Steps to Reproduce

1. set local variables
2. `terraform apply`

**Comment by favoretti** [comment]
Hi there, 

thank you for reporting this. We're using exactly the same config and I think the issue is here:
```
      user_assigned_identity_id = data.azurerm_user_assigned_identity.identity.user_assigned_identity_id
```

This should be just 
```
user_assigned_identity_id = data.azurerm_user_assigned_identity.identity.id
```

Can you please try this and let us know whether than fixes your issue?

**Comment by tehho** [comment]
my misstake with copying our modules.
We are using the id field of the azurerm_user_assigned_identity.

**Comment by tehho** [comment]
I've added support to our module for both service principal and user assigned identity with a dynamic block.
It shows the id of the user assigned identity in the logs and plan so not sure why it's not working.

```hcl
  dynamic "identity" {
    for_each = local.dynamic_identity
    content {
      type = identity.value.type
      user_assigned_identity_id = identity.value.user_assigned_identity_id
    }
  }

  dynamic "service_principal" {
    for_each = local.dynamic_service_principal
    content {
      client_id     = service_principal.value.client_id
      client_secret = service_principal.value.client_secret
    }
  }
```
where the locals are
```hcl
  dynamic_identity = var.identity == null ? {} : {
    identity = {
      type                      = "UserAssigned"
      user_assigned_identity_id = var.identity.id
    }
  }
  dynamic_service_principal = var.identity == null ? {
    service_principal = {
      client_id     = data.azurerm_key_vault_secret.cluster_client_id.value
      client_secret = data.azurerm_key_vault_secret.cluster_secret.value
    }
  } : {}
```
and var identity is
```hcl
variable "identity" {
  type = object({
    id = string
  })
  default = null
}
```

**Comment by tehho** [comment]
This script does the same thing. 
```bash
az group create -n Group-AKS-Lab
az identity create -n msi-aks-lab -g Group-AKS-Lab 
id=$(az identity show -n msi-aks-lab -g Group-AKS-Lab --query id -o json)
az aks create --name aks-lab-1 --resource-group Group-AKS-Lab --generate-ssh-key --enable-managed-identity --assign-identity "$id"
```

Have tried `az feature register --namespace Microsoft.ContainerService -n UserAssignedIdentityPreview` but no luck.

**Comment by favoretti** [comment]
This is a working config that we use with 2.50.. 

```
resource "azurerm_user_assigned_identity" "this" {
  name                = format("%s-identity", module.labels.id)
  resource_group_name = azurerm_resource_group.this.name
  location            = azurerm_resource_group.this.location
  tags                = module.labels.tags
}

resource "azurerm_role_assignment" "default_pool_subnet_route_table" {
  scope                = var.default_pool.vnet_subnet.route_table.id
  role_definition_name = "Contributor"
  principal_id         = azurerm_user_assigned_identity.this.principal_id
}

locals {
  subnets = { for pool_name, pool in var.node_pool_profiles :
    pool_name => pool.vnet_subnet.id
  }
}

resource "azurerm_kubernetes_cluster" "this" {
  depends_on              = [azurerm_role_assignment.default_pool_subnet_route_table]
  name                    = module.labels.id
  location                = azurerm_resource_group.this.location
  resource_group_name     = azurerm_resource_group.this.name
  node_resource_group     = format("%s-aks-rg", module.labels.id)
  dns_prefix              = module.labels.id
  kubernetes_version      = var.kubernetes_version
  private_cluster_enabled = var.private

  addon_profile {
    http_application_routing {
      enabled = var.enable_http_application_routing
    }

    kube_dashboard {
      enabled = false
    }

    dynamic "oms_agent" {
      for_each = var.install_oms_agent ? ["this"] : []
      content {
        enabled                    = var.enable_oms_agent
        log_analytics_workspace_id = local.log_analytics_workspace_id
      }
    }

    azure_policy {
      enabled = var.enable_azure_policy_addon
    }
  }

  default_node_pool {
    name                         = "system"
    node_count                   = var.default_pool.node_count
    min_count                    = var.default_pool.min_count
    max_count                    = var.default_pool.max_count
    max_pods                     = var.default_pool.max_pods
    vm_size                      = var.default_pool.vm_size
    orchestrator_version         = var.kubernetes_version
    enable_auto_scaling          = true
    only_critical_addons_enabled = true
    vnet_subnet_id               = var.default_pool.vnet_subnet.id
  }

  linux_profile {
    # Do not change this. AKS could be stuck in failed state if something goes wrong with username.
    admin_username = "azureuser"
    ssh_key {
      key_data = local.ssh_keys.public
    }
  }

  lifecycle {
    ignore_changes = [
      windows_profile,
      default_node_pool[0].node_count
    ]
  }

  network_profile {
    network_plugin     = var.network_plugin
    network_policy     = var.network_plugin == "azure" ? var.network_policy : null
    dns_service_ip     = var.dns_service_ip == null ? cidrhost(var.service_cidr, 10) : var.dns_service_ip
    docker_bridge_cidr = var.docker_bridge_cidr
    pod_cidr           = var.network_plugin == "kubenet" ? var.pod_cidr : null
    service_cidr       = var.service_cidr
    load_balancer_sku  = "Standard"

    load_balancer_profile {
      managed_outbound_ip_count = var.outbound_ip_count
    }
  }

  identity {
    type                      = "UserAssigned"
    user_assigned_identity_id = azurerm_user_assigned_identity.this.id
  }

  role_based_access_control {
    enabled = true

    azure_active_directory {
      managed                = true
      admin_group_object_ids = var.azure_ad_admin_groups
    }
  }

  tags = module.labels.tags
}
```

**Comment by tehho** [comment]
Tried this and still got the same problem.
Have opened a ticket with microsoft support and will update this issue with what they say.

**Comment by tehho** [comment]
Got it working!
TLDR: 
Aad-pod-identity along side kubelet_identity[0].user_assigned_identity_id needs owner over AKS resources resource group and reader over the userassigned identity.

When you set the userassigned identity for AKS you only replace the Azure side, not kubernetes side.
This means that you can set the identity if you want to control management of loadbalancers, public ips etc. 
But in order for you to control access to ACR or use the userassigned identity to managed other identitys you need to give the kubernetes identity owner over AKS resources resource group, reader over the userassigned identity and set up aad-pod-identity with a binding agains a dummy pod, in order for the identity to be bound to the cluster.
After all this kubernetes has access to the identity and can connect to the ACR.

Our concern right now is that we need to give our terraform service principal owner over the subscription since AKS creates a new resource group and we can not pregive user management rights to the service principal.
Was looking into PIM for access elevation but this is only in preview for programmatically elevate your access and PIM does not support service principals.

**Comment by justin-chizer** [comment]
@tehho Do you think there is a work around if we are not using Aad-pod-identity? I feel like I should be able to set the identityProfile.kubeletidentity to our user assigned identity.
